refactor documentation

christo4ferris opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
## Description

refactor of docs to make the use of fabric components more accessible to mere mortals and without requiring that binaries and images be built by the user. This turned into a major refactoring exercise. 
## Motivation and Context

<!-- Why is this change required? What problem does it solve? -->

<!-- If it fixes an open issue, please link to the issue here. -->

There were a number of things that needed to be addressed. First, there was lots of redundant documentation about how to start components etc that was either conflicting, or at the very least inconsistent. I attempted to refactor things so that there is setup, build, and operational separation. 

More could be and should be done, but I don't think it makes sense to leave this outstanding any longer.

Fixes #2144 
## How Has This Been Tested?

<!-- If this PR does not contain a new test case, explain why. -->

<!-- Describe in detail how you tested your changes. -->

I have gone through and tested the various directions personally.
## Checklist:

<!-- To check a box, and an 'x': [x] -->

<!-- To uncheck box, add a space: [ ] -->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff).
- [x] I have either added documentation to cover my changes or this change requires no new documentation.
- [x] I have either added unit tests to cover my changes or this change requires no new tests.
- [] I have run [golint](https://github.com/golang/lint) and have fixed valid warnings in code I have added or modified. This tool generates false positives so you may choose to ignore some warnings. The goal is clean, consistent, and readable code.

<!-- The continuous integration build process will run [make checks](https://github.com/hyperledger/fabric/blob/master/Makefile#L22) to confirm that tests pass and that code quality meets minimum standards. You may optionally run this locally as PRs will not be accepted until they pass. -->

Signed-off-by: Christopher Ferris chrisfer@us.ibm.com

christo4ferris committed:
Merge remote-tracking branch 'hyperledger/master'

christo4ferris committed:
Merge remote-tracking branch 'hyperledger/master'

christo4ferris committed:
initial refactor of devnet-setup

initial refactor of devnet-setup to improve the organization and to
make the use of fabric more accessible to mere mortals without
requiring a build of the binaries and images.

Fixes: #2144

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
I'm interested in initial feedback on my refactor of devnet-setup.md @joshhus @srderson @muralisrini 

I am also interested in help with membership services usage. What was previously described is far from comprehensive or informative. It probably deserves a document of its own. I don't feel confident that I know enough (yet) about use of membersrvc to be writing the tutorial, but am willing to give it a go with some guidance.

srderson commented on docs/dev-setup/devnet-setup.md:
```diff
@@ -1,31 +1,70 @@
-logging output from the `peer` and chaincodes.
+### Leveraging published Docker images
 
+This approach simply leverages the Docker images that the Hyperledger Fabric project publishes to [DockerHub](https://hub.docker.com/u/hyperledger/dashboard/) and either Docker commands or Docker Compose descriptions of the network one wishes to create.
```
Link is 404 for me. https://hub.docker.com/u/hyperledger works. Maybe because I'm not logged in?

srderson commented on docs/dev-setup/devnet-setup.md:
```diff
@@ -37,33 +76,126 @@ vagrant@vagrant-ubuntu-trusty-64:/opt/gopath/src/github.com/hyperledger/fabric$
+
+#### Start up a validating peer:
+
+Let's launch the first validating peer (the root node). We'll set CORE_PEER_ID to vp0 and CORE_VM_ENDPOINT asabove. Here's the docker-compose.yml for launching a single container within a **Vagrant** environment:
```
asabove -> as above

christo4ferris committed:
fix typo and 404

fix typo and broken link

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

srderson commented:
Looks better already. I'd suggest changing the name of the file from `devnet-setup.md` to `Setting Up a Network.md` as it will be more clear in the table of contents at http://hyperledger-fabric.readthedocs.io/en/latest/. Also, I would suggest moving out the the `dev-setup` folder to possibly just a `Setup` folder?

christo4ferris committed:
Merge remote-tracking branch 'hyperledger/master' into issue-2144

christo4ferris committed:
this is turning into a major refactor of docs

removed index.md from mkdocs.yml because it was a nothing burger
deleted index.md
created Setup directory
renamed SandboxSetup.md
Chaincode-setup.md and moved to Setup directory
renamed dev-setup/install.md build.md because it was really build
instructions
cleaned up build.md
moved dev-setup/JAVAChaincode.md to Setup
relevant changes in README.md to account for renaming of various files
removed TravisCI_Readme.md from mkdocs.yml because it is not relevant
to anyone but the Hyperledger CI team
moved TravisCI_Readme.md to top level directory
moved dev-setup/ca-setup.md to Setup
moved dev-setup/logging-control.md to Setup
made change in core.yml to account for relocated logging-control.md

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
@srderson took your suggestion and it is becoming a bit more involved than I had planned originally. However, I am pretty pleased thus far with the progress. I have moved all of the setup to a Setup directory and I added that to the main index. Then I had to accommodate all the moving around which triggered a bit more refactoring.

@rameshthoomu is there a way to generate the mkdocs against a pull request or a personal repo? I'd like to see how this looks.

christo4ferris committed:
the docs refactor continues

cleaned up ca-setup.md
use ‘Chaincode’ capitalization consistently in README
minor tweaks to devenv.md

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

rameshthoomu commented:
@christo4ferris : mkdocs works only on Push notifications. yes, it is possible to setup on a personal repo.
I have pulled this PR and did some modifications to mkdocs.yaml and included index.md in docs folder (It is required for mkdocs to build the documentation). Below is the mkdocs generated document looks like on this PR.

http://chaincode-docs.readthedocs.io/en/latest/. I have observed 404 not found errors in couple places. Will send you the list of observations in sometime. 

Just for reference, I have created draft version on ReadTheDocs Readme.md. https://github.com/rameshthoomu/fabric/blob/tools/docs/RTD.md

christo4ferris committed:
add links to TOC in protocol-specification.md

add links to TOC in protocol-specification.md

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris committed:
fix broken internal links

fix broken internal links

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
@rameshthoomu can you please investigate the jenkins "failure" - I used [ci skip] because just docs changed but Jenkins seems to think that skipping == fail. https://jenkins.hyperledger.org/job/fabric-github-verify-x86_64/557/console

rameshthoomu commented:
seems there is an open issue in Jenkins ci-skip plugin.. Created an issue in Jenkins JIRA.

christo4ferris committed:
work on node sdk doc

Merge remote-tracking branch 'hyperledger/master' into issue-2144
fix up node idk doc README.md

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

christo4ferris commented:
@hyperledger/fabric-maintainers @joshhus @nickgaski @mastersingh24 @smithbk I think that this is ready to merge. The docs still need some additional attention, but this hopefully makes things a bit better organized and hopefully up to date (worked on my machine;-)

christo4ferris commented:
check that... just had a discussion with @muralisrini about --peer-chaincodedev flag and I need to tweak the chaincode dev page a bit before we merge.

christo4ferris committed:
relocated and refactored the NodeSDK setup

relocated and refactored the NodeSDK setup
reworked some of the chaincode setup docs
other cosmetic tweaks

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
okay @hyperledger/fabric-maintainers @joshhus @nickgaski @mastersingh24 @smithbk I think that this is ready for review. Thanks.

christo4ferris commented:
Note that Jenkins failure is result of Jenkins bug with [ci skip]. There are no code changes in this PR (well, excepting to a comment in core.yaml that changed the reference to a markdown file).

christo4ferris committed:
fix broken link and mkdocs.yml

fix broken link and mkdocs.yml

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris committed:
more broken links fixed

more broken links fixed
minor cosmetic tweaks

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

srderson commented on sdk/node/README.md:
```diff
@@ -1,266 +1,20 @@
-* The [Going Deeper](#going-deeper) section discusses HFC's pluggability or extensibility design.  It also describes the main object hierarchy to help you get started in navigating the reference documentation. The top-level class is `Chain`.
-
-   WARNING: To view the reference documentation correctly, you first need to build the SDK and confirm functionality by [running the unit tests](#running-unit-tests).  Subsequently open the following URLs directly in your browser.  Be sure to replace YOUR-FABRIC-DIR with the path to your fabric directory.
+   **WARNING**: To view the reference documentation correctly, you first need to build the SDK and confirm functionality by [running the unit tests](#running-unit-tests).  Subsequently open the following URLs directly in your browser.  Be sure to replace YOUR-FABRIC-DIR with the path to your fabric directory.
```
Should be one space after periods instead of two.

srderson commented on sdk/node/README.md:
```diff
@@ -1,266 +1,20 @@
+To learn about how to install and use the Node.js SDK, please visit the [fabric documentation](http://hyperledger-fabric.readthedocs.io/en/latest/Setup/NodeSDK-setup).
 
-    npm install -g hfc
+The [Going Deeper](#going-deeper) section discusses HFC's pluggability or extensibility design.  It also describes the main object hierarchy to help you get started in navigating the reference documentation. The top-level class is `Chain`.
```
"pluggability or extensibility design" -> pluggability and extensibility design

Should be one space after periods instead of two.

srderson commented on docs/dev-setup/devenv.md:
````diff
@@ -49,22 +49,28 @@ git clone https://github.com/<username>/fabric.git
 ```
 
-**NOTE:** If you intend to run the development environment behind an HTTP Proxy, you need to configure the guest so that the provisioning process may complete.  You can achieve this via the *vagrant-proxyconf* plugin. Install with `vagrant plugin install vagrant-proxyconf` and then set the VAGRANT_HTTP_PROXY and VAGRANT_HTTPS_PROXY environment variables *before* you execute `vagrant up`. More det ...
+Go get coffee... this will take a few minutes. Once complete, you should be able to `ssh` into the Vagrant VM just created.
````
:)

srderson commented on docs/Setup/ca-setup.md:
```diff
@@ -0,0 +1,182 @@
+- `logging.error`
+- `logging.panic`
+
+Alternatively, these fields can be set via environment variables, which---if set---have precedence over entries in the yaml file.  The corresponding environment variables are named as follows:
```
Should be one space after periods

christo4ferris replied:
@srderson  this two spaces thing is not my editing/editor;-) happy to fix them, though.

srderson commented on docs/Setup/ca-setup.md:
```diff
@@ -0,0 +1,182 @@
+
+## Configuration
+
+All CA services are provided by a single process, which can be configured by setting parameters in the CA configuration file `membersrvc.yaml`, which is located in the same directory as the CA binary.  More specifically, the following parameters can be set:
```
Should be one space after periods

srderson commented on docs/Setup/ca-setup.md:
```diff
@@ -0,0 +1,182 @@
+
+### Transaction Certificate Authority
+
+Once a user is enrolled, he or she can also request _transaction certificates_ from the _transaction certificate authority_ (TCA). These certificates are to be used for deploying Chaincode and for invoking Chaincode transactions on the blockchain.  Although a single _transaction certificate_ can be used for multiple transactions, for privacy reasons it is recommended that a new _transaction certi ...
```
Should be one space after periods "blockchain.  Although"

christo4ferris committed:
address typos in sdk/node/README.md

address typos in sdk/node/README.md and docs/Setup/NodeSDK-setup.md

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

srderson commented on docs/Setup/Chaincode-setup.md:
```diff
@@ -1,70 +1,219 @@
+## Writing, Building, and Running Chaincode in a Development Environment
 
-Chaincode developers need a way to test and debug their chaincode without having to set up a complete peer network. This document describes how to write, build, and test chaincode in a local development environment.
+Chaincode developers need a way to test and debug their chaincode without having to set up a complete peer network. By default, when you want to interact with chaincode, you need to first `Deploy` it using the CLI, REST API or SDK. Upon receiving this request, the peer node would typically spin up a Docker container with the relevant chaincode. This can make things rather complicated for debuggin ...
```
"`Deploy` it using the CLI, REST API or SDK" -> "`Deploy` it using the CLI, REST API, gRPC API, or SDK"

srderson commented on docs/Setup/Chaincode-setup.md:
```diff
@@ -1,70 +1,219 @@
 
-    cd $GOPATH/src/github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02
-    go build
+To run the fabric-peer and fabric-membersrvc images, we'll use [Docker Compose](https://docs.docker.com/compose/). It significantly simplifies things. To do that, we''ll create a docker-compose.yml file with a description of the two services we'll be running. Here's the docker-compose.yml to launch the two processes:
```
we''ll -> we'll

JonathanLevi commented on docs/Setup/Chaincode-setup.md:
```diff
@@ -1,70 +1,219 @@
+- [Option 3](#Option-3-Docker-Toolbox) using Docker toolbox
 
-    vagrant ssh
+Using options *2* and *3* allow you to avoid having to build from and keep current a clone of the fabric GitHub repos.
```
3 suggestions: Using options 2 **and** 3 allow you to avoid having to **build from** and keep current a clone of the fabric GitHub **repos**.

How about:

Using option _2_ or _3_ will allow you to avoid having to build everything from scratch (along with the need to keep a clone of the fabric GitHub code current/up-to-date).

or (my preference, though I can't tell how this 'sounds'):

Using option 2 or 3 will preclude you from the need to build everything from scratch (and from keeping a clone of the fabric GitHub code current/up-to-date).

christo4ferris replied:
I revised but with a slightly different approach that I think factors in your concern.

srderson commented on docs/Setup/Chaincode-setup.md:
```diff
@@ -1,70 +1,219 @@
+
+### Running the Peer and CA
+
+To run the fabric-peer and fabric-membersrvc images, we'll use [Docker Compose](https://docs.docker.com/compose/). It significantly simplifies things. To do that, we''ll create a docker-compose.yml file with a description of the two services we'll be running. Here's the docker-compose.yml to launch the two processes:
```
we''ll -> we'll

christo4ferris committed:
fixed double space after period

fixed double space after period

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

JonathanLevi commented on docs/Setup/Chaincode-setup.md:
```diff
@@ -1,70 +1,219 @@
+  * [chaincode invoke via CLI and REST](#chaincode-invoke-via-cli-and-rest)
+  * [chaincode query via CLI and REST](#chaincode-query-via-cli-and-rest)
+
+If you were running with security enabled, see [Removing temporary files when security is enabled](#removing-temporary-files-when-security-is-enabled) to learn how to clean up the temp files.
```
temp files -> temporary files

christo4ferris committed:
fix we''ll typo

fix we''ll typo

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

srderson commented:
@christo4ferris Looks great! Added some minor comments.

Hopefully others can review soon so this can be merged in time for the upcoming hackfest.

JonathanLevi commented on docs/Setup/Network-setup.md:
```diff
@@ -0,0 +1,261 @@
+#### Pulling the images from DockerHub
+
+Once you have Docker (1.11 or greater) installed and running, and
+before starting any of the fabric components, you will need to first pull the fabric images from DockerHub.
```
Once you have Docker (1.11 or greater) installed and running, [and before] starting any of the fabric components, you will need to first pull the fabric images from DockerHub.

to:

Once you have Docker (1.11 or greater) installed and running, **prior to** starting any of the fabric components, you will need to first pull the fabric images from DockerHub.

christo4ferris committed:
address review comments

address review comments

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
@srderson @JonathanLevi thanks for the review! I've addressed the comments (I think).

JonathanLevi commented on docs/Setup/Network-setup.md:
```diff
@@ -0,0 +1,261 @@
+
+**Note:** When running Docker _natively_ on Mac and Windows, there is no IP forwarding support available. Hence, running more than one fabric-peer image is not advised because you do not want to have multiple processes binding to the same port. For most application and chaincode development/testing running with a single fabric peer should not be an issue unless you are interested in performance a ...
+
+With this approach, there are multiple choices as to how to run Docker. Using [Docker Toolbox](https://docs.docker.com/toolbox/overview/) or one of the new native Docker runtime environments for [Mac OSX](https://docs.docker.com/engine/installation/mac/) or [Windows](https://docs.docker.com/engine/installation/windows/). There are some subtle differences between how Docker runs natively on Mac an ...
```
where appropriate, below when we get to

to:

where appropriate below, when we get to

[that is, move the comma to after `below`]

JonathanLevi commented on docs/tech/application-ACL.md:
````diff
@@ -118,50 +118,50 @@ message Transaction {
 ```
 
-Another way to achieve this is to have the payload containing the metadata itself.  
+Another way to achieve this is to have the payload containing the metadata itself.
````
have the payload **contain** the meta...

JonathanLevi commented on docs/tech/application-ACL.md:
```diff
@@ -118,50 +118,50 @@ message Transaction {
 
-To assist chaincode execution, the validators provide the chaincode additional information, like the metadata and the binding.  
- 
+To assist chaincode execution, the validators provide the chaincode additional information, like the metadata and the binding.
```
like -> such as

JonathanLevi commented on docs/tech/application-ACL.md:
```diff
@@ -118,50 +118,50 @@ message Transaction {
-that privilege or role. The latter is done offline. 
 
-Now, Alice requires that to invoke the *hello* function, a certain message *M* has to be authenticated by an authorized invoker (Bob, in our case). 
+Alice has full control on the deployment transaction's metadata.
```
full control **over** the deployment...

JonathanLevi commented on docs/tech/application-ACL.md:
```diff
@@ -118,50 +118,50 @@ message Transaction {
-Now, Alice requires that to invoke the *hello* function, a certain message *M* has to be authenticated by an authorized invoker (Bob, in our case). 
+Alice has full control on the deployment transaction's metadata.
+In particular, the metadata can be used to store a list of ACLs (one per function), or a list of roles.
+To define each of these lists/roles, Alice can use any TCerts/ECerts of the users who have been assigned
```
To define each of these lists/roles, Alice can use any TCerts/ECerts of the users who have been assigned that privilege or role. 

to 

To define each of these lists/roles, Alice can use any TCerts/ECerts of the users who have been 
**granted that (access control) privilege or have been assigned that role**

JonathanLevi commented on docs/tech/application-ACL.md:
```diff
@@ -118,50 +118,50 @@ message Transaction {
+To define each of these lists/roles, Alice can use any TCerts/ECerts of the users who have been assigned
+that privilege or role. The latter is done offline.
+
+Now, Alice requires that to invoke the *hello* function, a certain message *M* has to be authenticated by an authorized invoker (Bob, in our case).
```
Alice requires that **in order** to invoke the _hello_ function

JonathanLevi commented on docs/tech/application-ACL.md:
```diff
@@ -118,50 +118,50 @@ message Transaction {
 
 1. Bob gets a CertificateHandler for CertBob, *cHandlerBob*;
-2. Bob gets a new TransactionHandler to issue the execute transaction, *txHandler* relative to his next available TCert or his ECert; 
+2. Bob gets a new TransactionHandler to issue the execute transaction, *txHandler* relative to his next available TCert or his ECert;
```
gets -> obtains
- [ minor, but do we want some `TransactionHandler` style/formatting? minor, as in: really minor! ]

christo4ferris committed:
address additional review comments

address additional review comments

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris committed:
address grammar issue

address grammar issue

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris committed:
add SystemChaincode/noop.md

add SystemChaincode/noop.md

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
@nickgaski @joshhus @mastersingh24 @smithbk would greatly appreciate a review, thanks! Would like to have this version available prior to the hackfest.

JonathanLevi commented on docs/tech/attributes.md:
```diff
@@ -5,7 +5,7 @@ Attributes values are encrypted using the keys defined below (section Attributes
 ## Attributes Keys
 
-The attributes are encrypted using a key derived from a hierarchy called PreKey tree. This approach consists in deriving keys from a parent key, allowing the parent key owner, get access to derived keys. This way keys used to encrypt attributes are different among attributes and TCerts avoiding linkability while allowing an authorized auditor who owns a parent key to derive the keys in the lower  ...
+The attributes are encrypted using a key derived from a hierarchy called PreKey tree. This approach consists in deriving keys from a parent key, allowing the parent key owner, get access to derived keys. This way keys used to encrypt attributes are different among attributes and TCerts avoiding linkability while allowing an authorized auditor who owns a parent key to derive the keys in the lower  ...
```
Attributes are encrypted using a key derived from a key hierarchy, denoted as the `PreKey tree`. 

The `Pre-Key tree` key hierarchy provides a powerful pattern for generating and using cryptographic keys, in which keys at a lower level of the tree are derived from their parent/master key. In general, such key hierarchy allows one to use different keys for different data.

In this context, the `PreKey tree` provides us with the ability to use different keys to encrypt different attributes (even within the same `TCert`). This is useful both for addressing potential transaction linkability concerns and for simplifying key management.

Note that the `PreKey tree` allows an authorized auditor (who owns a parent/master key) to derive the keys further down in the key hierarchy without requiring further communication with other entities.

christo4ferris replied:
All I did was remove a trailing space;-) Do you want to submit a PR to my branch?

JonathanLevi replied:
Yes, that's true - sorry, I'm on a roll here ;-) Please ignore it and I'll update separately. We'd have better history this way.

JonathanLevi commented on docs/tech/attributes.md:
```diff
@@ -5,7 +5,7 @@ Attributes values are encrypted using the keys defined below (section Attributes
 
 ## Attributes Keys
```
Attribute Keys

[a single, Attribute]

JonathanLevi commented on docs/tech/attributes.md:
```diff
@@ -140,10 +140,10 @@ Attribute Certificate Authority (ACA) has the responsibility of certify the owne
-4.  TCA could have other criteria to update the valid list of attributes.
-5.  TCA creates the batch of TCerts. Each TCert contains the valid attributes encrypted with keys derived from the Prekey tree (each key is unique per attribute, per TCert and per user).
-6.  TCA returns the batch of TCerts to the user along with a root key (Prek0) from which each attribute encryption key was derived. There is a Prek0 per TCert. All the TCerts in the batch have the same attributes and the validity period of the TCerts is the same for the entire batch.
+3. TCA checks the validity period of the ACert's attributes and updates the list by eliminating those that are expired. Then for scenarios b and c from the previous item it checks how many (and which ones) of the attributes the user will actually receive inside each TCert. This information needs to be returned to the user in order to decide whether the TCerts are useful or if further actions need ...
```
The TCA

JonathanLevi commented on docs/tech/attributes.md:
```diff
@@ -140,10 +140,10 @@ Attribute Certificate Authority (ACA) has the responsibility of certify the owne
-5.  TCA creates the batch of TCerts. Each TCert contains the valid attributes encrypted with keys derived from the Prekey tree (each key is unique per attribute, per TCert and per user).
-6.  TCA returns the batch of TCerts to the user along with a root key (Prek0) from which each attribute encryption key was derived. There is a Prek0 per TCert. All the TCerts in the batch have the same attributes and the validity period of the TCerts is the same for the entire batch.
+3. TCA checks the validity period of the ACert's attributes and updates the list by eliminating those that are expired. Then for scenarios b and c from the previous item it checks how many (and which ones) of the attributes the user will actually receive inside each TCert. This information needs to be returned to the user in order to decide whether the TCerts are useful or if further actions need ...
+4. TCA could have other criteria to update the valid list of attributes.
```
The TCA

JonathanLevi commented on docs/tech/attributes.md:
```diff
@@ -140,10 +140,10 @@ Attribute Certificate Authority (ACA) has the responsibility of certify the owne
-6.  TCA returns the batch of TCerts to the user along with a root key (Prek0) from which each attribute encryption key was derived. There is a Prek0 per TCert. All the TCerts in the batch have the same attributes and the validity period of the TCerts is the same for the entire batch.
+3. TCA checks the validity period of the ACert's attributes and updates the list by eliminating those that are expired. Then for scenarios b and c from the previous item it checks how many (and which ones) of the attributes the user will actually receive inside each TCert. This information needs to be returned to the user in order to decide whether the TCerts are useful or if further actions need ...
+4. TCA could have other criteria to update the valid list of attributes.
+5. TCA creates the batch of TCerts. Each TCert contains the valid attributes encrypted with keys derived from the Prekey tree (each key is unique per attribute, per TCert and per user).
```
The TCA

JonathanLevi commented on docs/tech/attributes.md:
```diff
@@ -140,10 +140,10 @@ Attribute Certificate Authority (ACA) has the responsibility of certify the owne
+3. TCA checks the validity period of the ACert's attributes and updates the list by eliminating those that are expired. Then for scenarios b and c from the previous item it checks how many (and which ones) of the attributes the user will actually receive inside each TCert. This information needs to be returned to the user in order to decide whether the TCerts are useful or if further actions need ...
+4. TCA could have other criteria to update the valid list of attributes.
+5. TCA creates the batch of TCerts. Each TCert contains the valid attributes encrypted with keys derived from the Prekey tree (each key is unique per attribute, per TCert and per user).
+6. TCA returns the batch of TCerts to the user along with a root key (Prek0) from which each attribute encryption key was derived. There is a Prek0 per TCert. All the TCerts in the batch have the same attributes and the validity period of the TCerts is the same for the entire batch.
```
The TCA

JonathanLevi commented:
@christo4ferris OK, I'm done reviewing this for today. It just goes to show that I don't read the manuals (often enough). 

Sorry for the bursty traffic earlier. There's (always!) more to do on the Wiki, but it is pretty orthogonal to this refactoring work/PR.  Will probably be better to merge this work in sooner.

christo4ferris committed:
fix grammar nits

fix grammar nits

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

joshhus commented:
Will do ... are we talking about everything under "Dev setup" at http://hyperledger-fabric.readthedocs.io/en/latest/ ? ... (when is the hackfest) ...

joshhus commented:
Is there an editor to handle minor tweaks. Remove duplicate period, e.g., etc., in CA Setup.  Such as, "with root-signed ones.." ... and: 
it creates all its required state (e.g., internal databases, CA certificates, blockchain keys, etc.) and write  >>
it creates all its required <states> (e.g., internal databases, CA certificates, blockchain keys, etc.) and <writes> ...

joshhus commented:
Otherwise looking good ... A prospective improvement might be a "plain English" explanation of what each function does. e.g. for CA, at the top under Overview, something like: "The CA provides trusted interactions for blockchain users. The CA controls who has access to the blockchain network, who can submit transactions, and a secure communication protocol for all transactions."

christo4ferris committed:
fix typos

fix typos

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
@joshhus actually, you can see all my changes in the Files Changed tab above.

joshhus commented:
Right, that's faster okay. Should I just post tweaks here? 
e.g. MemberServicesAPI.md: "If the user has not **be** >> **been** registered before,"

(I'd also suggest eventually moving towards the "singular their" -- to avoid "his" and "her" gendered references.) 
e.g. " ... by specifiying **her** >> **their** name and roles ..."

joshhus commented:
(note *\* bold *\* and >> or << chars indicates the suggested change and the direction)
(Seems "indeed" should be deleted 2x, unless it's a special term of art.)
MemberServicesAPI.md: 
The `CreateCertificatePair` **functions >> function** allows a user to create and read **her >> their** enrollment certificate pair. For this, the user has to do two successive invocations of this **functions >> function**. Firstly, both the signature and encryption public keys have to be handed to the ECA together with the one-time password **previously** << returned by the `RegisterUser` function invocation << **before**. The request has to be signed by the user's private signature key to demonstrate that the user is in possession of the private signature key << **indeed**. The ECA in return gives the user a challenge encrypted with the user's public encryption key. The **user** has to decrypt the challenge, thereby demonstrating that **she is >> they are** in possession of the private encryption key << **indeed**, and re-issue the certificate creation request passing **<< to pass** the decrypted challenge instead of the one-time password passed in the invocation. If the challenge has been decrypted correctly, the ECA issues and returns the enrollment certificate pair for the user.

joshhus commented:
Chaincode-setup.md: 
(seems "process" should be deleted)
"If you are enabling security and privacy on the peer << **process** with environment variables, "

joshhus commented:
Chaincode-setup.md:
Is the .yml extension correct for Docker / we use .yaml. Multiple instances, e.g. of "docker-compose.yml" 

typo: 
"temporary stop-gap until we can provide an **indepdendent << independent** package for the chaincode shim."

joshhus commented:
JAVAChaincode.md:
1. Follow the step << **steps** 1 to 3 as above,

christo4ferris commented:
@joshhus it is much easier for the person who submits a PR to deal with in-line comments. If you click on the Files Changed tab, green is new content added in the pull request. Red is deleted content. Click on a (green) line opens up a comment window and you can leave your comment there. Makes it easier to find the code/text to change. Also allows a threaded discussion if needed. 

as for .yaml vs .yml yaml is more correct but for docker compose the [default](https://docs.docker.com/compose/gettingstarted/#/step-3-define-services) is expecting a file named docker-compose.yml.

christo4ferris committed:
address review feedback from @joshhus

address review feedback from @joshhus

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

joshhus commented:
Network-setup.md:
"**.... how to run Docker. Using " >> ... how to run Docker: using ...**  (2nd clause not independent)

"If you are using the native Docker for Mac or Windows, the value for `CORE_VM_ENDPOINT` should be set to `unix:///var/run/docker.sock`. \ **[TODO] double check this. I believe that `127.0.0.1:2375` also works.**"

"If we wanted to use the docker commandline " << *\* Docker ? << command line **

<!-- This needs to be sorted out with a revamped security section ...  **TBA**

*\* " to use Byzantine consensus plugin >> to use the Practical Byzantine Fault Tolerance consensus plugin ." **

"All of these setting may be overriden << **overridden** via the command line environment variables, eg.  <<  *\* e.g.  **"

" controlling logging output from the `peer` and chaincodes." << "... 'peer' and **deployed chaincode** ... ?? 

<!--
+**Note:** When running  **TBA**

"The Hyperledger Fabric Client SDK (HFC)" << *\* The Hyperledger Fabric Client (HFC) SDK **

joshhus commented:
NodeSDK-setup.md:
"... some of it's common capabilities. " << " ... some of **its** common ... "

"shows you how to setup up your ..." << "... to **set up** your ... "

"to demonstrate all of it's power, " << " ... all of **its** power ... "

joshhus commented:
okay will move these to inline comments. when is the hackathon?

joshhus commented on docs/API/MemberServicesAPI.md:
```diff
@@ -13,7 +13,7 @@ The administrator interface of the ECA provides the following functions:
 	}
 
-The `RegisterUser` function allows you to register a new user by specifiying her name and roles in the `RegisterUserReq` structure.  If the user has not be registered before, the ECA registers the new user and returns a unique one-time password, which can be used by the user to request her enrollment certificate pair via the public interface of the ECA.  Otherwise an error is returned.
+The `RegisterUser` function allows you to register a new user by specifiying their name and roles in the `RegisterUserReq` structure. If the user has not be registered before, the ECA registers the new user and returns a unique one-time password, which can be used by the user to request their enrollment certificate pair via the public interface of the ECA. Otherwise an error is returned.
```
... If the user has not "been" registered ...

joshhus commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+
+The sections in this document are as follows:
+
+* The [Getting Started](#getting-started) section is intended to help you quickly get a feel for HFC, how to use it, and some of it's common capabilities. This is demonstrated by example.
```
"... its common ..."

joshhus commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+## Hyperledger Fabric Client SDK for Node.js
+
```
" ... Hyperledger Fabric Client (HFC) SDK ... "

joshhus commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+## Hyperledger Fabric Client SDK for Node.js
+
+The Hyperledger Fabric Client SDK (HFC) provides a powerful and easy to use API to interact with a Hyperledger Fabric blockchain.
```
"The Hyperledger Fabric Client (HFC) SDK ... "

joshhus commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+
+* The [Getting Started](#getting-started) section is intended to help you quickly get a feel for HFC, how to use it, and some of it's common capabilities. This is demonstrated by example.
+
+* The [Getting Set Up](#getting-set-up) section shows you how to setup up your environment and how to [run the unit tests](#running-unit-tests). Looking at the implementation of the unit tests will also help you learn more about the APIs by example, including asset management and confidentiality.
```
"... how to set up ..."

joshhus commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+* The [Getting Set Up](#getting-set-up) section shows you how to setup up your environment and how to [run the unit tests](#running-unit-tests). Looking at the implementation of the unit tests will also help you learn more about the APIs by example, including asset management and confidentiality.
+
+## Getting Started
+This purpose of this section is to help you quickly get a feel for HFC and how you may use it. It is not intended to demonstrate all of it's power, but to demonstrate common use cases by example.
```
" ... its power ..."

joshhus commented on docs/Setup/Network-setup.md:
```diff
@@ -0,0 +1,261 @@
+-->
+
+### Using a Consensus Plugin
+A consensus plugin might require some specific configuration that you need to set up. For example, to use Byzantine consensus plugin provided as part of the fabric, perform the following configuration:
```
"... to use the Practical Byzantine Fault Tolerance (PBFT) consensus protocol, which is provided as ... "

joshhus commented on docs/Setup/Network-setup.md:
```diff
@@ -0,0 +1,261 @@
+
+See `core.yaml` and `consensus/pbft/config.yaml` for more detail.
+
+All of these setting may be overriden via the command line environment variables, eg. `CORE_PEER_VALIDATOR_CONSENSUS_PLUGIN=pbft` or `CORE_PBFT_GENERAL_MODE=batch`
```
"... overridden ..."
"e.g."

JonathanLevi replied:
"e.g., "

joshhus commented on docs/Setup/Network-setup.md:
```diff
@@ -0,0 +1,261 @@
+### Logging control
+
+See [Logging Control](logging-control.md) for information on controlling
+logging output from the `peer` and chaincodes.
```
" 'peer' and deployed chaincode."  ?

joshhus commented on docs/Setup/Network-setup.md:
````diff
@@ -0,0 +1,261 @@
+    - vp0
+```
+
+If we wanted to use the docker commandline to launch another peer, we need to get the IP address of the first validating peer, which will act as the root node to which the new peer(s) will connect. The address is printed out on the terminal window of the first peer (e.g. 172.17.0.2) and should be passed in with the `CORE_PEER_DISCOVERY_ROOTNODE` environment variable.
````
"... Docker command line ... "

joshhus commented on docs/Setup/Network-setup.md:
```diff
@@ -0,0 +1,261 @@
+
+**Note:** When running Docker _natively_ on Mac and Windows, there is no IP forwarding support available. Hence, running more than one fabric-peer image is not advised because you do not want to have multiple processes binding to the same port. For most application and chaincode development/testing running with a single fabric peer should not be an issue unless you are interested in performance a ...
+
+With this approach, there are multiple choices as to how to run Docker. Using [Docker Toolbox](https://docs.docker.com/toolbox/overview/) or one of the new native Docker runtime environments for [Mac OSX](https://docs.docker.com/engine/installation/mac/) or [Windows](https://docs.docker.com/engine/installation/windows/). There are some subtle differences between how Docker runs natively on Mac an ...
```
"... Docker: using ... "

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+
+### Chaincode Deployment Directory Structure
+
+To have the chaincode deployment succeed in network mode, you must properly set up the chaincode project outside of your Hyperledger Fabric source tree. These instructions will demonstrate how to properly set up the directory structure to deploy *chaincode_example02* in network mode.
```
These instructions >> The below instructions

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+
+To have the chaincode deployment succeed in network mode, you must properly set up the chaincode project outside of your Hyperledger Fabric source tree. These instructions will demonstrate how to properly set up the directory structure to deploy *chaincode_example02* in network mode.
+
+The chaincode project must be placed inside the `src` directory in your local `$GOPATH`. For example, the [chaincode_example02](https://github.com/hyperledger/fabric/blob/master/examples/chaincode/go/chaincode_example02/chaincode_example02.go) project may be placed inside `$GOPATH/src/` as shown below.
```
**under** the `src` directory

...
may be placed **under** `$GOPATH/src/`

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
````diff
@@ -0,0 +1,253 @@
+curl GET https://raw.githubusercontent.com/hyperledger/fabric/master/examples/chaincode/go/chaincode_example02/chaincode_example02.go > chaincode_example02.go
+```
+
+Once you have placed your chaincode project inside the `src` directory in your local `$GOPATH`, you will need to vendor the dependencies. In your chaincode source directory, run the following commands:
````
**After placing** your code **under** the `src` directory...

JonathanLevi replied:
**From the directory containing your chaincode source, execute** the following commands:

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
````diff
@@ -0,0 +1,253 @@
+var testChaincodePath = "github.com/chaincode_example02/";
+```
+
+**Note:** You will need to run `npm install` the first time you run the sdk tests, to install all of the dependencies. Set the `DEPLOY_MODE` environment variable to `net` and run the chain-tests as follows:
````
**in order to install** ...

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+
+### Enabling TLS
+
+If you wish to activate TLS connection with the member services the following actions are needed:
```
If you wish to configure TLS with the Membership Services server, the following steps are required:

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
````diff
@@ -0,0 +1,253 @@
+            file: "/var/hyperledger/production/.membersrvc/tlsca.priv"
+```
+
+This is needed to instruct the member services on which tls cert and key to use.
````
Specifying to the Membership Services (TLS) Certificate Authority (TLSCA) what X.509 v3 Certificate (with a  corresponding Private Key) to use.

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
````diff
@@ -0,0 +1,253 @@
+                file: "/var/hyperledger/production/.membersrvc/tlsca.cert"
+```
+
+This is needed to allow the peer to connect to the member services using TLS, otherwise the connection will fail.
````
Configuring the peer to connect to the Membership Services server over TLS (otherwise, the connection will fail).

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+
+This is needed to allow the peer to connect to the member services using TLS, otherwise the connection will fail.
+
+- Bootstrap your member services and the peer. This is needed in order to have the file *tlsca.cert* generated by the member services.
```
member services >>Membership Services

JonathanLevi replied:
member services >>Membership Services

(twice)

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+*Note:* If you cleanup the folder `/var/hyperledger/production` then don't forget to copy again the *tlsca.cert* file as described above.
+
+## Running Unit Tests
+HLC includes a set of unit tests implemented with the [tape framework](https://github.com/substack/tape). The unit [test script](https://github.com/hyperledger/fabric/blob/master/sdk/node/bin/run-unit-tests.sh) builds and runs both the membership service server and the peer node for you, therefore you do not have to start those manually.
```
membership services >>Membership Services

The unit [test script] >> The [unit test script]

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
```diff
@@ -0,0 +1,253 @@
+The [asset-mgmt-with-roles.js](https://github.com/hyperledger/fabric/blob/master/sdk/node/test/unit/asset-mgmt-with-roles.js) test case exercises the [asset_management_with_roles.go](https://github.com/hyperledger/fabric/tree/master/examples/chaincode/go/asset_management_with_roles) chaincode when it has been deployed in both development mode and network mode.
+
+#### Troublingshooting
+If you see errors stating that the client has already been registered/enrolled, keep in mind that you can perform the enrollment process only once, as the enrollmentSecret is a one-time-use password. You will see these errors if you have performed a user registration/enrollment and subsequently deleted the crypto tokens stored on the client side. The next time you try to enroll, errors similar to ...
```
crypto >> cryptographic

JonathanLevi commented on docs/Setup/NodeSDK-setup.md:
````diff
@@ -0,0 +1,253 @@
+   Error: user is already registered
+   ```
+
+To address this, remove any stored crypto material from the CA server by following the instructions [here](https://github.com/hyperledger/fabric/blob/master/docs/API/SandboxSetup.md#removing-temporary-files-when-security-is-enabled). You will also need to remove any of the crypto tokens stored on the client side by deleting the KeyValStore directory. That directory is configurable and is set to ` ...
````
cryptographic tokens

JonathanLevi commented on docs/Setup/ca-setup.md:
```diff
@@ -0,0 +1,182 @@
+## Certificate Authority (CA) Setup
```
I will deal with this entire `ca-setup` file separately, in a separate PR!

christo4ferris replied:
please do!

christo4ferris committed:
address more reviewer feedback

address more reviewer feedback

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris committed:
address remaining reviewer comments

address remaining reviewer comments

Signed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris committed:
one last one?

one last one?

Sigbed-off-by: Christopher Ferris <chrisfer@us.ibm.com>

[ci skip]

christo4ferris commented:
thanks @srderson @JonathanLevi @joshhus for the reviews!

srderson commented:
LGTM. Thanks for the reviews @JonathanLevi and @joshhus ! We can continue making improvements in future PRs.

joshhus commented:
I'm adding comments to the PR, much better method.

joshhus commented:
ah okay, next one.

joshhus commented on docs/Setup/ca-setup.md:
```diff
@@ -0,0 +1,182 @@
+
+For example, a peer that is also a validator would have a role value of 6.
+
+When the CA is started for the first time, it will generate all of its required state (e.g., internal databases, CA certificates, blockchain keys, etc.) and writes this state to the directory given in its configuration. The certificates for the CA services (i.e., for the ECA, TCA, and TLSCA) are self-signed as the current default. If those certificates shall be signed by some root CA, this can be ...
```
"... of its required state attributes ..."  ? or "states" ... 
"... and write its state ..." or "states"

christo4ferris replied:
@joshhus sorry, thought you were done... @JonathanLevi plans to take on ca-setup.md. Please feel free to submit PRs for any other changes.

joshhus commented:
I looked at some of the other files incl. /devenv.md and /build.md, and find nothing major - looks good. Minor nits to look out for moving fwd (can reduce translation flags down the road):
- Vagrant and Docker and Behave always capitalized, unless as dir names, e.g.
- When introducing a following code sample or list, e.g., end the lead-in sentence with a colon: (rather than nothing, or a period.)
- Use "set up" two words when it's a verb; use "setup" one word when it's a noun.
